### ISSUE TITLE
Improve NagiosXI authenticated exploit modules to increase resilience and for use with Autocheck disabled

### ISSUE THREAD

**Issue by k0pak4** [rapid7/metasploit-framework#17606]
## Summary

During the course of https://github.com/rapid7/metasploit-framework/pull/17494 there were several concerns with how the NagiosXI login mixin was being used in the check method, which prevents the module from working when Autocheck is disabled. Additionally, other improvements were made including cleaner regexes in version detection, nil checks on objects that were assumed to be not nil, and other improvements. After examining the other NagiosXI modules the following modules should be modified to also take advantage of these improvements:

- modules\exploits\linux\http\nagios_xi_autodiscovery_webshell.rb
- modules\exploits\linux\http\nagios_xi_mibs_authenticated_rce.rb
- modules\exploits\linux\http\nagios_xi_plugins_check_plugin_authenticated_rce.rb
- modules\exploits\linux\http\nagios_xi_plugins_filename_authenticated_rce.rb

### Improvements

- [x] Move the authentication to its own function, and call it in check and exploit when necessary (https://github.com/rapid7/metasploit-framework/pull/17494#discussion_r1092327656)
- [x] Use better regexes for version detection (https://github.com/rapid7/metasploit-framework/pull/17494#discussion_r1097807544)
- [x] Refactor case statements to use accurate error codes, error messages, and Failure codes (https://github.com/rapid7/metasploit-framework/pull/17494#discussion_r1096464021)
- [ ] Improve documentation to give detailed installation instructions (see documentation\modules\exploit\linux\http\nagios_xi_configwizards_authenticated_rce.md)
- [ ] Improve lib\msf\core\exploit\remote\http\nagios_xi\login.rb to make the return structure consistent (currently reversed on failure from success) (https://github.com/rapid7/metasploit-framework/pull/17494#discussion_r1097985116)

Generally, these modules can also be cleaned up and shortened similarly to how the config wizards RCE module was through PR.

### Motivation

Currently, these modules will fail with AutoCheck disabled, so we want to improve that first and foremost. Additionally, the version checking and error codes will provide more support when running the modules against old versions of NagiosXI.

## Vulnerable Software
In general, older versions of NagiosXI can be found:
- As an OVA (replace version number in the link to the one you want to download): https://assets.nagios.com/downloads/nagiosxi/5/ovf/nagiosxi-5.7.5-64.ova
- As a package: https://www.nagios.com/downloads/nagios-xi/older-releases/

documentation\modules\exploit\linux\http\nagios_xi_configwizards_authenticated_rce.md has detailed installation instructions if help is needed on install

**Comment by Sidharthareddy99** [comment]
I want to work on this issue please assign me this issue.

**Comment by gwillcox-r7** [comment]
@Sidharthareddy99 Your assigned now, let me know if you need additional info on anything.

**Comment by k0pak4** [comment]
Yup feel free to ask questions if you need @Sidharthareddy99 , it's very top of mind at the moment 😅

**Comment by gwillcox-r7** [comment]
Unassigning for now, if a PR is put up though I can reassign. We mostly use assignment internally to track who is working on what.  Anyone is free to put up a PR to fix an issue at any time, at which point we can look into assigning people at that point

**Comment by xehil** [comment]
Can I get to work on this? I'd like to contribute.

**Comment by gwillcox-r7** [comment]
Sure @xehil, feel free to take it on 👍 Do you need help forking the repo? I see you haven't done so yet.

**Comment by xehil** [comment]
@gwillcox-r7 thank you. I was waiting for the issue to as assigned.

**Comment by gwillcox-r7** [comment]
https://github.com/rapid7/metasploit-framework/pull/17820 resolves most of these issues however the final two points still stand. Reopening this issue as they will need to be addressed before we can consider this truly fixed.

**Comment by AlejandroCantu** [comment]
Hi @gwillcox-r7 I am a student at UT Austin in the Ethical Hacking class who would like to contribute on this issue as a final project for the class. I see the final two points stand, but upon reading [this comment](https://github.com/rapid7/metasploit-framework/pull/17494#discussion_r1098009805), it seems there is deeper structural changes and checks to pass if the order of the return values of `visit_nagios_dashboard` change? I'm not sure I fully understanding the issue because the place where `visit_nagios_dashboard` is called does not use the return values. Could you help me understand where these return values from `visit_nagios_dashboard` are used?

**Comment by gwillcox-r7** [comment]
> Hi @gwillcox-r7 I am a student at UT Austin in the Ethical Hacking class who would like to contribute on this issue as a final project for the class. I see the final two points stand, but upon reading [this comment](https://github.com/rapid7/metasploit-framework/pull/17494#discussion_r1098009805), it seems there is deeper structural changes and checks to pass if the order of the return values of `visit_nagios_dashboard` change? I'm not sure I fully understand the issue because the place where `visit_nagios_dashboard` is called does not use the return values. Could you help me understand where these return values from `visit_nagios_dashboard` are used?

So as for the issue I think the main problem is seen at https://github.com/rapid7/metasploit-framework/blob/f6c8181b7f3edc3f29f41a323c2f8eb054619cd5/lib/msf/core/exploit/remote/http/nagios_xi/login.rb#L251-L257 where we can see that for one response the response cookies come first and then the response body and yet in another response the response body comes first and then the cookies.

Ideally we should be using hashes here as well instead of arrays to prevent issues related to positional issues. However we decide to do this though, all calling code will need to be updated to account for the changes that we make.

As for where the return codes from `visit_nagios_dashboard` are used, you will want to see where it is called from first. Sourcegraph is a great tool to help with this. https://sourcegraph.com/github.com/rapid7/metasploit-framework with `context:global repo:^github\.com/rapid7/metasploit-framework$` is a good start and then we can search for `visit_nagios_dashboard` there.

Looking for this shows one reference at https://sourcegraph.com/github.com/rapid7/metasploit-framework/-/blob/lib/msf/core/exploit/remote/http/nagios_xi/login.rb?L180 which is in the `nagios_xi_login` function. Click on that function name and then "View References" and you should see about 15 references from where that is called.

Hopefully that helps :)

**Comment by AlejandroCantu** [comment]
Grant, I've got a solution that standardizes the authentications in the NagiosXI Scanner and one of the older exploits to not directly call `nagios_xi_login`, but use `authenticate`, like `nagios_xi_configwizards_authenticated_rce.rb`. 

Firstly, I changed the order of the return from `visit_nagios_dashboard` and changed `authenticate` accordingly. Also made a small change to `extract_auth_cookies` for the new format, which doesn't break anything else since this is the only call to `extract_auth_cookies`.

**Comment by k0pak4** [comment]
@nrathaus it looks like the last two requirements are not resolved as far as I can tell? The PR addressing the last one was closed with an attic tag, and from what I see the response codes still are sent in the [same format](https://github.com/rapid7/metasploit-framework/blob/71ce80730095506a532fde58976cb4b2818a67c4/lib/msf/core/exploit/remote/http/nagios_xi/login.rb#L254C1-L257C40)? Unless I'm missing something I don't think the last requirement is addressed

**Comment by nrathaus** [comment]
Previous versions of NagiosXI are listed here: https://assets.nagios.com/downloads/nagiosxi/versions.php

**Comment by k0pak4** [comment]
@nrathaus yes the previous versions are also listed in the original description, but I don't think this is fully addressed
